README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# tactileR: R package for creating tactile graphics for users with visual impairments.

[![License: GPL
v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)
<!-- badges: start --> [![Travis build
status](https://travis-ci.org/jooyoungseo/tactileR.svg?branch=master)](https://travis-ci.org/jooyoungseo/tactileR)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/jooyoungseo/tactileR?branch=master&svg=true)](https://ci.appveyor.com/project/jooyoungseo/tactileR)
[![Codecov test
coverage](https://codecov.io/gh/jooyoungseo/tactileR/branch/master/graph/badge.svg)](https://codecov.io/gh/jooyoungseo/tactileR?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/tactileR)](https://cran.r-project.org/package=tactileR)
<!-- badges: end -->

## Description

Just surround any graphic object between two functions `brl_begin()` and
`brl_end()`. The graph will be created in PDF format in your current
directory, with all text being presented using a braille font. See more
details in the example section below.

## Installation

### Development Version

You can install the latest development version as follows:

``` r
if(!require(devtools)) {
install.packages("devtools")
}

devtools::install_github('jooyoungseo/tactileR')
```

### Stable Version

You can install the released version of tactileR from
[CRAN](https://CRAN.R-project.org)
with:

``` r
install.packages("tactileR")
```

## Arguments

| Argument |                                                                                                           Description                                                                                                           |
| :------- | :-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| file     |                                                   A character string giving the filename where the image is to be saved. The default file name is “tactile.pdf” if not given.                                                   |
| pt       |                                                                          Numeric value specifying braille font size for a created file. Default is 11.                                                                          |
| paper    |                                                                          Character value setting paper size of output pdf file. Default is “special”.                                                                           |
| font     | Character value for font family. Available options include “BRL” for “BRAILLE1.ttf” (Braille Normal) and “SWL” for “Swell-Braille.ttf” (Swell Braille), and “TIMES” for “times.ttf” (Times New Roman). Default is set to “BRL”. |

## Details

*Note*: This function optimizes for both 11*11 and 11*8.5 swell touch
papers. When printing a generated braille pdf, just remember two things:

1.  check “Fit” in the print dialog box (the shortcut key is ALT+F).
2.  Check “Landscape” in the print dialog box as well.

<!-- end list -->

  - For boxplot, you do not need to check the “Landscape,” just leave it
    as a “Portray.” However, please check the “Fit” all the time.

## Information

  - If you are not familiar with what swell form machine is, please
    refer to [this web
    site](http://www.americanthermoform.com/product/swell-form-graphics-ii-machine/).
  - A short [YouTube video](https://www.youtube.com/watch?v=ClI555l4Z1M)
    will help you to understand how this function can be employed in
    conjunction with Swell form machine.

## Return

Nothing within the R session, but a pdf file will be created in the
user’s working directory.

## Examples

### Histogram

``` r
# Loading tactile package:
library(tactileR)

# Create tactile graphic by surrounding between two functions like below. The arguments used in this example are all given by default:
brl_begin(file='tactile.pdf', pt=11, paper='special', font='BRL')
hist(airquality$Ozone)
brl_end()
# The braille ready-to-print pdf version will be created in your working directory. Print it out and run the swell touch paper through swell form heating machine.
```

### Boxplot

``` r
library(tactileR)

x <- rnorm(100)
brl_begin('test.pdf')
boxplot(x)
brl_end()
# The braille ready-to-print pdf version will be created in your working directory. Print it out and run the swell touch paper through swell form heating machine.
```

*Note*: Histogram and boxplot are just examples. You can feed any
graphic object through the functions. barplot, interaction plot,
regression plot, network graph, wordcloud, etc.

## Limitation

Currently this package does not support braille translation. This means
you cannot embed any special symbols, math symbols, contracted braille,
and non-roman alphabet characters. This issue will be addressed later by
employing Liblouis translator.


NEWS.md
# tactileR 0.1.0

* Initial release.


tests/testthat.R
library(testthat)
library(tactileR)

test_check("tactileR")


tests/testthat/test-brl_end.R
context("test-brl_end")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


tests/testthat/test-brl_begin.R
context("test-brl_begin")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


R/frame.R

frame <- function(tmp='article', pkg='tactileR') {

message("\nEnter your project name.")
file <- readLines(n = 1)
if (file == "") {file = "new project"}

if(tmp != 'week') {
message("\nEnter the title of your paper.")
title <- readLines(n = 1)
if (title == "") {title = "Untitled"}
}

if(tmp == 'article') {
message("\nEnter the running head.")
short <- readLines(n = 1)
if (short == "") {short = "runninghead"}
}

if(tmp %in% c('article', 'pt', 'reaction', 'week', 'html')) {
rmarkdown::draft(paste0(file, '.Rmd'), template=tmp, package=pkg, create_dir=T, edit=F)
} else {
rmarkdown::draft(paste0(file, '.Rmd'), template=tmp, package=pkg, create_dir=T, edit=T)
}

if(tmp != 'week') {
titleFile <- file(paste0(file, '/title.txt'))
writeLines(title, titleFile)
close(titleFile)
}
if(tmp == 'article') {
write(toupper(short), paste0(file, '/title.txt'), append=T)
}

if(tmp == 'article') {
fileConn <- file(paste0(file, '/run.R'))
writeLines(c(
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), output_format="papaja::apa6_word")'), 
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), output_format="papaja::apa6_pdf")'), 
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), output_format="html_document")'), 
 paste0('browseURL(normalizePath("', file, '.docx"))'), 
paste0('browseURL(normalizePath("', file, '.pdf"))'),
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), output_format="powerpoint_presentation")') 
), fileConn)
close(fileConn)
file.edit(paste0(file, '/', 'main.txt'))
} else if(tmp %in% c('reaction', 'week', 'pt', 'html')) {
fileConn <- file(paste0(file, '/run.R'))
writeLines(c(
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), rmarkdown::word_document(toc=T, toc_depth=3, highlight="tango", reference_docx = system.file("rmarkdown/templates/apa6/resources/apa6_man.docx", package="papaja")))'), 
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), output_format="powerpoint_presentation")'), 
paste0('rmarkdown::render(normalizePath("', file, '.Rmd"), output_format="pdf_document")'), 
paste0('browseURL(normalizePath("', file, '.docx"))'), 
paste0('browseURL(normalizePath("', file, '.pdf"))'), 
paste0('rmarkdown::render(normalizePath("', file, '.Rmd")')
), fileConn)
close(fileConn)
file.edit(paste0(file, '/', 'main.txt'))
} else {
fileConn <- file(paste0(file, '/run.R'))
writeLines(c(paste0('rmarkdown::render(normalizePath("', file, '.Rmd"))'), 
paste0('browseURL(normalizePath("', file, '.pptx"))'), 
paste0('browseURL(normalizePath("', file, '.html"))'), 
paste0('browseURL(normalizePath("', file, '.docx"))'), 
paste0('browseURL(normalizePath("', file, '.pdf"))')), fileConn)
close(fileConn)
}

browseURL(normalizePath(paste0(file, '/bib/references.bib')))

# Google Scholar
scholar <- as.logical(readline("\nDo you want to open Google Scholar? (No)\n"))
if (is.na(scholar)) scholar <- FALSE
if (scholar) {
browseURL('http://scholar.google.com')
}

readline(
"Press <enter> once the viewing and editing is completed. \nThen, the source file will be processed.")

if(tmp == 'article') {
sink(paste0(file, '/log.txt'))
rmarkdown::render(paste0(file, '/', file, '.Rmd'), output_format="papaja::apa6_word")
sink()
} else if(tmp %in% c("reaction", "week")) {

sink(paste0(file, '/log.txt'))
rmarkdown::render(paste0(file, '/', file, '.Rmd'), rmarkdown::word_document(toc=T, toc_depth=3, highlight="tango", reference_docx = system.file("rmarkdown/templates/apa6/resources/apa6_man.docx", package="papaja")))
sink()

} else {
sink(paste0(file, '/log.txt'))
rmarkdown::render(paste0(file, '/', file, '.Rmd'))
sink()
}


readline(
"Press <enter> to view the output file.")

if(file.exists(paste0(file, "/", file, ".docx"))) {browseURL(normalizePath(paste0(file, "/", file, ".docx")))}
else if(file.exists(paste0(file, "/", file, ".pptx"))) {browseURL(normalizePath(paste0(file, "/", file, ".pptx")))}
else if(file.exists(paste0(file, "/", file, ".html"))) {browseURL(normalizePath(paste0(file, "/", file, ".html")))}

readline(
"Press <enter> to working directory of the current project.")

browseURL(normalizePath(file))

if(tmp == 'article') {

pdfOK <- as.logical(readline("\nDo you want to make pdf file? (No)\n"))
if (is.na(pdfOK)) pdfOK <- FALSE
if (pdfOK) {
rmarkdown::render(normalizePath(paste0(file, '/', file, '.Rmd')), output_format='papaja::apa6_pdf')
if(file.exists(paste0(file, "/", file, ".pdf"))) {browseURL(normalizePath(paste0(file, "/", file, ".pdf")))}
}

}
}



R/BRL.end.r
#' BRL.end
#' @title Generate and optimize a braille-embedded ready-to-print pdf file for swell form machine or braille embosser

#' @aliases BRL.end
#' @keywords BRL.end

#' @description Closing braille device for your graphic. Just surround any graphic object between two functions BRL.begin() and BRL.end(). The graph will be created in PDF format in your current directory, with all text being presented using  a braille font. See more details in the example section below.

#' @export BRL.end

#' @details
#' *Note*: This function optimizes for both 11*11 and 11*8.5 swell touch papers. When printing a generated braille pdf, just remember two things:
#' 1. check "Fit" in the print dialog box (the shortcut key is ALT+F).
#' 2. Check "Landscape" in the print dialog box as well.
#' * For boxplot, you do not need to check the "Landscape," just leave it as a "Portray." However, please check the "Fit" all the time.

#' Information:
#' If you are not familiar with what swell form machine is, please refer to [this web site](http://www.americanthermoform.com/product/swell-form-graphics-ii-machine/).
#' A short Youtube video will help you to understand how this function can be employed in conjunction with Swell form machine: https://www.youtube.com/watch?v=ClI555l4Z1M

#' @return Nothing within the R session, but a pdf file will be created in the user's working directory.

#' @examples

#' ```{r, echo=TRUE}
#' BRL.begin(file='tactile.pdf', pt=11, paper='special', font='BRL')
#' hist(airquality$Ozone)
#' BRL.end()
#' # The braille ready-to-print pdf version will be created in your working directory. Print it out and run the swell touch paper through swell form heating machine.
#' ```

#' @author JooYoung Seo

#Function starts
BRL.end <-
function() {

# Body code Starts here
showtext::showtext.end()
dev.off()

# Body code ends
}
#Function Ends


R/brl_end.R
#' brl_end
#' @title Generate and optimize a braille-embedded ready-to-print pdf file for swell form machine or braille embosser

#' @aliases brl_end
#' @keywords brl_end

#' @description Closing braille device for your graphic. Just surround any graphic object between two functions brl_begin() and brl_end(). The graph will be created in PDF format in your current directory, with all text being presented using  a braille font. See more details in the example section below.

#' @export brl_end

#' @details
#' *Note*: This function optimizes for both 11*11 and 11*8.5 swell touch papers. When printing a generated braille pdf, just remember two things:
#' 1. check "Fit" in the print dialog box (the shortcut key is ALT+F).
#' 2. Check "Landscape" in the print dialog box as well.
#' * For boxplot, you do not need to check the "Landscape," just leave it as a "Portray." However, please check the "Fit" all the time.

#' Information:
#' If you are not familiar with what swell form machine is, please refer to \href{http://www.americanthermoform.com/product/swell-form-graphics-ii-machine/}{this web site}.
#' A short \href{https://www.youtube.com/watch?v=ClI555l4Z1M}{YouTube video} will help you to understand how this function can be employed in conjunction with Swell form machine.

#' @return Nothing within the R session, but a pdf file will be created in the user's working directory.

#' @examples
#' brl_begin(file='tactile.pdf', pt=11, paper='special', font='BRL')
#' hist(airquality$Ozone)
#' brl_end()
#' # The braille ready-to-print pdf version will be created in your working directory.
#' # Print it out and run the swell touch paper through swell form heating machine.

#' @author JooYoung Seo, \email{jooyoung@psu.edu}

#Function starts
brl_end <-
function() {

# Body code Starts here
showtext::showtext.end()
grDevices::dev.off()

# Body code ends
}
#Function Ends


R/BRL.begin.r
#' BRL.begin
#' @title Generate and optimize a braille-embedded ready-to-print pdf file for swell form machine or braille embosser

#' @aliases BRL.begin
#' @keywords BRL.begin

#' @description Opening braille device for your graphic. Just surround any graphic object between two functions BRL.begin() and BRL.end(). The graph will be created in PDF format in your current directory, with all text being presented using  a braille font. See more details in the example section below.

#' @export BRL.begin
#' @param file A character string giving the filename where the image is to be saved. The default file name is "tactile.pdf" if not given.
#' @param pt Numeric value specifying braille font size for a created file. Default is 11.
#' @param paper Character value setting paper size of output pdf file. Default is "special".
#' @param font Character value for font family. Available options include "BRL" for "BRAILLE1.ttf" (Braille Normal) and "SWL" for "Swell-Braille.ttf" (Swell Braille), and "TIMES" for "times.ttf" (Times New Roman). Default is set to "BRL".

#' @details
#' *Note*: This function optimizes for both 11*11 and 11*8.5 swell touch papers. When printing a generated braille pdf, just remember two things:
#' 1. check "Fit" in the print dialog box (the shortcut key is ALT+F).
#' 2. Check "Landscape" in the print dialog box as well.
#' * For boxplot, you do not need to check the "Landscape," just leave it as a "Portray." However, please check the "Fit" all the time.

#' Information:
#' If you are not familiar with what swell form machine is, please refer to [this web site](http://www.americanthermoform.com/product/swell-form-graphics-ii-machine/).
#' A short Youtube video will help you to understand how this function can be employed in conjunction with Swell form machine: https://www.youtube.com/watch?v=ClI555l4Z1M

#' @return Nothing within the R session, but a pdf file will be created in the user's working directory.

#' @examples

#' ```{r, echo=TRUE}
#' BRL.begin(file='tactile.pdf', pt=11, paper='special', font='BRL')
#' hist(airquality$Ozone)
#' BRL.end()
#' # The braille ready-to-print pdf version will be created in your working directory. Print it out and run the swell touch paper through swell form heating machine.
#' ```

#' @author JooYoung Seo

#Function starts
BRL.begin <-
function(file="tactile.pdf", pt=11, paper="special", font="BRL") {

# Body code Starts here
#Preparing braille font
sysfonts::font.add("BRL", system.file("fonts/BRAILLE1.ttf", package="tactileR"))
sysfonts::font.add("SWL", system.file("fonts/Swell-Braille.ttf", package="tactileR"))
sysfonts::font.add("TIMES", system.file("fonts/times.ttf", package="tactileR"))

#Generating Braille PDF
pdf(file, pointsize=pt, paper=paper)
showtext::showtext.begin()
#Optimizing graphic size
par(cex=1.2, cex.main=1.2, cex.lab=1.2, cex.axis=1.2, cex.sub=1.2, mar=c(5,5,3,2), pch=20, family=font)

# Body code ends
}
#Function Ends


R/brl_begin.R
#' brl_begin
#' @title Generate and optimize a braille-embedded ready-to-print pdf file for swell form machine or braille embosser

#' @aliases brl_begin
#' @keywords brl_begin

#' @description Opening braille device for your graphic. Just surround any graphic object between two functions brl_begin() and brl_end(). The graph will be created in PDF format in your current directory, with all text being presented using  a braille font. See more details in the example section below.

#' @export brl_begin
#' @param file A character string giving the filename where the image is to be saved. The default file name is "tactile.pdf" if not given.
#' @param pt Numeric value specifying braille font size for a created file. Default is 11.
#' @param paper Character value setting paper size of output pdf file. Default is "special".
#' @param font Character value for font family. Available options include "BRL" for "BRAILLE1.ttf" (Braille Normal) and "SWL" for "Swell-Braille.ttf" (Swell Braille), and "TIMES" for "times.ttf" (Times New Roman). Default is set to "BRL".

#' @details
#' *Note*: This function optimizes for both 11*11 and 11*8.5 swell touch papers. When printing a generated braille pdf, just remember two things:
#' 1. check "Fit" in the print dialog box (the shortcut key is ALT+F).
#' 2. Check "Landscape" in the print dialog box as well.
#' * For boxplot, you do not need to check the "Landscape," just leave it as a "Portray." However, please check the "Fit" all the time.

#' Information:
#' If you are not familiar with what swell form machine is, please refer to \href{http://www.americanthermoform.com/product/swell-form-graphics-ii-machine/}{this web site}.
#' A short \href{https://www.youtube.com/watch?v=ClI555l4Z1M}{YouTube video} will help you to understand how this function can be employed in conjunction with Swell form machine.

#' @return Nothing within the R session, but a pdf file will be created in the user's working directory.

#' @examples
#' brl_begin(file='tactile.pdf', pt=11, paper='special', font='BRL')
#' hist(airquality$Ozone)
#' brl_end()
#' # The braille ready-to-print pdf version will be created in your working directory.
#' # Print it out and run the swell touch paper through swell form heating machine.

#' @author JooYoung Seo, \email{jooyoung@psu.edu}

#Function starts
brl_begin <-
function(file="tactile.pdf", pt=11, paper="special", font="BRL") {

# Body code Starts here
#Preparing braille font
sysfonts::font.add("BRL", system.file("fonts/BRAILLE1.ttf", package="tactileR"))
sysfonts::font.add("SWL", system.file("fonts/Swell-Braille.ttf", package="tactileR"))
sysfonts::font.add("TIMES", system.file("fonts/times.ttf", package="tactileR"))

#Generating Braille PDF
grDevices::pdf(file, pointsize=pt, paper=paper)
showtext::showtext.begin()
#Optimizing graphic size
graphics::par(cex=1.2, cex.main=1.2, cex.lab=1.2, cex.axis=1.2, cex.sub=1.2, mar=c(5,5,3,2), pch=20, family=font)

# Body code ends
}
#Function Ends
